### PR TITLE
Support PullRequest builds from other repos

### PR DESCRIPTION
**Pull request by AdamBrousseau**
- Determine if build is a PR
- Automatically checkout PR and any dependent PRs
- Support dependent changes from other repos

[ci skip]

Issue #660

Signed-off-by: Adam Brousseau <adam.brousseau88@gmail.com>

#### TODO
- [x] Update README
- [x] Open Eclipse [Bugzilla](https://bugs.eclipse.org/bugs/show_bug.cgi?id=532195) to request Jenkins Webhook be added to openj9-omr repo

### CODE REVIEW

**Comment by AdamBrousseau** [comment]
adam test sanity plinux

**Comment by AdamBrousseau** [comment]
adam test sanity plinux jdk9

**Comment by AdamBrousseau** [comment]
adam test sanity plinux jdk9

File: buildenv/jenkins/common/build [hunk 1]
```diff
@@ -50,102 +50,154 @@ def get_source() {
...
+            omr_bool = true
+            break
+        case "eclipse/openj9":
+            openj9_bool = true
+            break
+        case ~/ibmruntimes(.*)/:
+            openjdk_bool = true
+            break
+        default:
+            error "Unable to determine sourece repo for PullRequest"
```
Comment (DanHeidinga) [inline buildenv/jenkins/common/build]: `sourece` -> `source`

**Comment by AdamBrousseau** [comment]
adam test sanity plinux jdk9

**Comment by AdamBrousseau** [comment]
adam test sanity plinux jdk9

**Comment by AdamBrousseau** [comment]
adam test sanity plinux jdk9

**Comment by AdamBrousseau** [comment]
adam test sanity plinux jdk9

**Comment by AdamBrousseau** [comment]
adam test sanity plinux jdk9

**Comment by AdamBrousseau** [comment]
adam test sanity plinux jdk9

**Comment by AdamBrousseau** [comment]
adam test sanity plinux jdk9 depends eclipse/omr#2357 ibmruntimes/openjdk-jdk8#58

**Comment by AdamBrousseau** [comment]
adam test sanity plinux jdk9 depends eclipse/omr#2357 ibmruntimes/openj9-openjdk-jdk8#58

**Comment by AdamBrousseau** [comment]
[Regular OpenJ9 PR](https://ci.eclipse.org/openj9/view/Sandbox/job/tmp_PullRequest-Sanity-JDK9-linux_ppc-64_cmprssptrs_le-OpenJ9/19/console)
[OpenJ9 PR with 2 dependent changes](https://ci.eclipse.org/openj9/view/Sandbox/job/tmp_PullRequest-Sanity-JDK9-linux_ppc-64_cmprssptrs_le-OpenJ9/21/console)
[OpenJDK8 PR with dependent 2 changes](https://ci.eclipse.org/openj9/view/Sandbox/job/tmp_PullRequest-Sanity-JDK9-linux_ppc-64_cmprssptrs_le-OpenJ9/23/console)

Do we want to protect against people specifying dependent changes from the incorrect version of an openjdk-extensions repo?
eg. `jenkins test sanity all jdk9 depends ibmruntimes/openj9-openjdk-jdk8#123`

**Comment by pshipton** [comment]
> Do we want to protect against people specifying dependent changes from the incorrect version of an openjdk-extensions repo?

I think so. Better than ignoring the dependent change and running the build anyway. How would it work? Perhaps running a job that immediately fails with an explanation of the problem.

**Comment by AdamBrousseau** [comment]
adam test sanity plinux jdk9 depends eclipse/omr#2357 ibmruntimes/openj9-openjdk-jdk8#58

**Comment by AdamBrousseau** [comment]
adam test sanity plinux jdk9 depends eclipse/omr#2357 ibmruntimes/openj9-openjdk-jdk8#58

**Comment by AdamBrousseau** [comment]
adam test sanity plinux jdk9 depends eclipse/omr#2357 ibmruntimes/openj9-openjdk-jdk9#108

**Comment by AdamBrousseau** [comment]
[Dependent JDK change with correct SDK version](https://ci.eclipse.org/openj9/view/Sandbox/job/tmp_PullRequest-Sanity-JDK9-linux_ppc-64_cmprssptrs_le-OpenJ9/24/console)
[Dependent JDK change with wrong SDK version](https://ci.eclipse.org/openj9/view/Sandbox/job/tmp_PullRequest-Sanity-JDK9-linux_ppc-64_cmprssptrs_le-OpenJ9/25/console)

**Comment by AdamBrousseau** [comment]
@vsebe for review

**Comment by vsebe** [comment]
desk review

**Comment by AdamBrousseau** [comment]
updated

**Comment by AdamBrousseau** [comment]
adam test sanity plinux jdk9 depends eclipse/omr#2357

**Comment by AdamBrousseau** [comment]
recent adjustment to switch-case still works
https://ci.eclipse.org/openj9/view/Sandbox/job/tmp_PullRequest-Sanity-JDK9-linux_ppc-64_cmprssptrs_le-OpenJ9/26/console

**Comment by AdamBrousseau** [comment]
updated

**Comment by vsebe** [comment]
LGTM

**Comment by AdamBrousseau** [comment]
Ready for committer review @pshipton or @DanHeidinga

**Comment by keithc-ca** [comment]
I'm not convinced this is a good direction to go. Rather than adjust the infrastructure to manage pull requests that must be coordinated across multiple repositories, I think we should aim to have pull requests that can be merged sequentially (first omr then openj9, or vice-versa). It may require a little more thought and work on the part of the author of the PRs, but I am confident that those PRs can always be constructed so there is an order they can be merged without build breakage at any intermediate state.

It can also get very complicated: consider a PR that is connected to openj9-openjdk changes. Currently this PR will only let one specify dependencies on one extension repo. If changes are required for jdk8, it's likely it also affects jdk9 (and jdk10, etc. - something not supported by the current proposal): where do we draw the line? I say that line should be at zero dependencies.

**Comment by AdamBrousseau** [comment]
@keithc-ca I agree that not having/needing dependent changes is a good idea. However, this PR is not about adding support for dependent changes, it is simply extending PR builds to other repos, which in turn extends the dependent change support. Also, there is an open issue (#827) to extend the "depends" support to also work with branches rather than just open PRs.

**Comment by AdamBrousseau** [comment]
If we want to discuss the validity of having dependent changes can we open a separate issue where the team leads can discuss? I don't think that conversation should be a blocker to this PR.
@keithc-ca @DanHeidinga

**Comment by keithc-ca** [comment]
On the contrary, I don't think we should be complicating things (like merging this PR would) until we have agreed that we want to consider PRs that come with dependencies.

**Comment by AdamBrousseau** [comment]
Keep in mind, we already support having pull requests with dependent changes. The purpose of this PR is extending PR Build functionality to the extensions repos. The only added functionality is the protection against cross-version dependent changes, which you could argue should be in a different PR as we actually have this vulnerability today, without these changes.

**Comment by keithc-ca** [comment]
I still don't think we should be doing this. If (and that a big *if*) we want to consider dependent PRs involving the extension repos, then it needs to allow specification of *all* the extension repos, not at most one.

**Comment by AdamBrousseau** [comment]
https://github.com/eclipse/openj9/blob/master/buildenv/jenkins/README.md#dependent-changes

**Comment by pshipton** [comment]
There are times when having dependent changes is just easier. Perhaps they could be committed in multiple steps to avoid the dependencies, but why would we want to do that dance when the changes can be committed (very closely) together. Seems like a waste of developer, committer, and machine time to avoid a very small timing hole. Dependent omr changes can be problematic because of the omr acceptance build, but there is no such issue for committing dependencies between openj9 and the extensions repos. 

With this PR, builds for all the extensions repos can be done, it just requires multiple commands. This could be improved in the future, if there is nothing better to do, but its not a requirement that needs to block merging this.

**Comment by DanHeidinga** [comment]
@pshipton I agree with you this is the right approach for now but have to take issue with:

> but why would we want to do that dance when the changes can be committed (very closely) together. Seems like a waste of developer, committer, and machine time to avoid a very small timing hole. 

That small timing hole leaves the repos in a broken state.  It stops tools like `git bisect` from running.  It leaves other developers with potentially broken repos.

Dependent changes externalize the cost of delivering a change from the dev making the change to everyone consuming the project.  It's not a great trade off.

Unfortunately, there are some dependent changes that we can't avoid.  We need this as a tool for those (rare) cases.  

Committers should continue to encourage all developers to stage their changes so we don't need this kind of tool.

File: buildenv/jenkins/README.md [hunk 2]
```diff
@@ -72,11 +74,9 @@ You can also request a "Compile & Sanity" build from the extensions repos or ope
 - Ex. Dependent change in OpenJDK Pull Request `#789`
     - `Jenkins test sanity depends ibmruntimes/openj9-openjdk-jdk9#789`
```
Comment (DanHeidinga) [inline buildenv/jenkins/README.md]: Isn't this line missing `all jdk9` as well?

File: buildenv/jenkins/README.md [hunk 3]
```diff
@@ -72,11 +74,9 @@ You can also request a "Compile & Sanity" build from the extensions repos or ope
...
 - Ex. Dependent changes in OMR and OpenJDK
-    - `Jenkins test sanity depends eclipse/omr#123 ibmruntimes/openj9-openjdk-jdk9#789`
+    - `Jenkins test sanity all jdk9 depends eclipse/omr#123 ibmruntimes/openj9-openjdk-jdk9#789`
```
Comment (DanHeidinga) [inline buildenv/jenkins/README.md]: This fixes one of these - aren't other commands incorrect as well?

**Comment by pshipton** [comment]
@DanHeidinga agreed. Good point, the timing hole I mentioned can sometimes be avoided by committing the changes in the correct order, even if they are within a few seconds of each other.

**Comment by keithc-ca** [comment]
I hadn't thought of the use cases where it simply makes life easier for a committer - those make sense. Ultimately, a committer has the responsibility to ensure that merges go smoothly and if this helps, I'm fine with it.

**Pull request merged**